File: schemas/loyalty-rewards-program/drizzle/member_activities.ts
// member_activities: Log of member actions that may trigger earning rules.
// See README.md for full design rationale.

import { pgTable, uuid, text, integer, jsonb, timestamp, index } from "drizzle-orm/pg-core";
import { loyaltyMembers } from "./loyalty_members";
import { pointsTransactions } from "./points_transactions";

export const memberActivities = pgTable("member_activities", {
    id: uuid("id").primaryKey().defaultRandom(),
    memberId: uuid("member_id").notNull().references(() => loyaltyMembers.id, { onDelete: "cascade" }),
    activityType: text("activity_type").notNull(),
    description: text("description"),
    source: text("source"),
    referenceType: text("reference_type"),
    referenceId: text("reference_id"),
    monetaryValue: integer("monetary_value"),
    currency: text("currency"),
    pointsAwarded: integer("points_awarded"),
    transactionId: uuid("transaction_id").references(() => pointsTransactions.id, { onDelete: "set null" }),
    metadata: jsonb("metadata"),
    createdAt: timestamp("created_at", { withTimezone: true }).notNull().defaultNow(),
  }, (table) => [
    index("idx_member_activities_member_id_created_at").on(table.memberId, table.createdAt),
    index("idx_member_activities_activity_type").on(table.activityType),
    index("idx_member_activities_reference").on(table.referenceType, table.referenceId),
    index("idx_member_activities_transaction_id").on(table.transactionId),
  ]
);

File: schemas/loyalty-rewards-program/spacetimedb/member_activities.rs
// member_activities: Log of member actions that may trigger earning rules.
// See README.md for full design rationale.
use spacetimedb::Timestamp;

#[spacetimedb::table(name = member_activities, public)]
pub struct MemberActivity {
    #[primary_key]
    pub id: String, // UUID

    #[index(btree)]
    pub member_id: String, // UUID, FK → loyalty_members.id (cascade delete)
    // Composite index: (member_id, created_at) — enforce in reducer logic
    #[index(btree)]
    pub activity_type: String,
    pub description: Option<String>,
    pub source: Option<String>,
    #[index(btree)]
    pub reference_type: Option<String>,
    // Composite index: (reference_type, reference_id) — enforce in reducer logic
    pub reference_id: Option<String>,
    pub monetary_value: Option<i32>,
    pub currency: Option<String>,
    pub points_awarded: Option<i32>,
    #[index(btree)]
    pub transaction_id: Option<String>, // UUID, FK → points_transactions.id (set null)
    pub metadata: Option<String>, // JSON stored as string
    pub created_at: Timestamp,
}
